L1Trigger/TrackerTFP/src/DataFormats.cc

```diff
@@ -0,0 +1,752 @@
+  }
+
+  // construct TrackKFin from TTTrackRef
+  TrackKFin::TrackKFin(const TTTrackRef& ttTrackRef, const DataFormats* dataFormats, const TTBV& hitPattern, const TTBV& layerMap, const TTBV& maybePattern, double phiT, double qOverPt, double zT, double cot, int sectorPhi, int sectorEta, int trackId) :
+    Track(ttTrackRef, dataFormats, Process::kfin, hitPattern, layerMap, maybePattern, phiT, qOverPt, zT, cot, sectorPhi, sectorEta, trackId),
+    stubs_(setup()->numLayers())
+  {
+    dataFormats_->convertTrack(data_, frame_.second, p_);
+  }
+
+  vector<TTStubRef> TrackKFin::ttStubRefs(const TTBV& hitPattern, const vector<int>& layerMap) const {
+    vector<TTStubRef> stubs;
+    stubs.reserve(hitPattern.count());
+    for (int layer = 0; layer < setup()->numLayers(); layer++)
+      if (hitPattern[layer])
+        stubs.push_back(stubs_[layer][layerMap[layer]]->ttStubRef());
+    return stubs;
+  }
+
+  // construct TrackKF from Frame
+  TrackKF::TrackKF(const FrameTrack& frame, const DataFormats* dataFormats) :
+    Track(frame, dataFormats, Process::kf) {}
+
+  // construct TrackKF from TrackKfin
+  TrackKF::TrackKF(const TrackKFin& track, double phiT, double qOverPt, double zT, double cot) :
+    Track(track, track.phiT(), track.qOverPt(), track.zT(), track.cot(), track.sectorPhi(), track.sectorEta(), false)
+  {
+    get<0>(data_) += phiT;
+    get<1>(data_) += qOverPt;
+    get<2>(data_) += zT;
+    get<3>(data_) += cot;
+    get<6>(data_) = abs(qOverPt) < dataFormats_->format(Variable::qOverPt, Process::sf).base() / 2. && abs(phiT) < dataFormats_->format(Variable::phiT, Process::sf).base() / 2.;
+    dataFormats_->convertTrack(data_, frame_.second, p_);
+  }
+
+  // conversion to TTTrack with given stubs
+  TTTrack<Ref_Phase2TrackerDigi_> TrackKF::ttTrack(const vector<StubKF>& stubs) const {
+    TTBV hitVector(0, setup()->numLayers());
+    double chi2phi(0.);
+    double chi2z(0.);
+    vector<TTStubRef> ttStubRefs;
+    const int nLayer = stubs.size();
+    ttStubRefs.reserve(nLayer);
+    for (const StubKF& stub : stubs) {
+      hitVector.set(stub.layer());
+      const TTStubRef& ttStubRef = stub.ttStubRef();
+      chi2phi += pow(stub.phi(), 2) / setup()->v0(ttStubRef, this->qOverPt());
+      chi2z += pow(stub.z(), 2) / setup()->v1(ttStubRef, this->qOverPt());
+      ttStubRefs.push_back(ttStubRef);
+    }
+    static constexpr int nParPhi = 2;
+    static constexpr int nParZ = 2;
+    static constexpr int nPar = nParPhi + nParZ;
+    const int dofPhi = nLayer - nParPhi;
+    const int dofZ = nLayer - nParZ;
+    chi2phi /= dofPhi;
+    chi2z /= dofZ;
+    const int sectorPhi = frame_.first->phiSector();
+    const int sectorEta = frame_.first->etaSector();
+    const double qOverPt = this->qOverPt() / setup()->invPtToDphi();
+    const double phi0 = deltaPhi(this->phiT() + this->qOverPt() * dataFormats_->chosenRofPhi() + dataFormats_->format(Variable::phiT, Process::ht).range() * (sectorPhi - .5));
```

Comment:
I am not sure I understand L454. Is this converting phiT to a global phi0 that is wanted by TTTrack ? If so, I thought the part about phi sector offset should be simply ( sector index) * (delta phi of each phi sector). When I printed the phi0 that is calculated by L454 and compared it with frame_.first->momentum().phi(), they look very different ?

Reply:
It is converting phiT which is track phi at radius T w.r.t. phi sector center into track phi at radius 0 w.r.t. nonant center. Please recognize that phi sectors are not nonants (we do split each nonant in two further phi sectors).